File: src/test/java/ru/privatenull/pnlibrary/update/SemanticVersionTest.java
package ru.privatenull.pnlibrary.update;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class SemanticVersionTest {

    @Test
    void acceptsPrefixAndShortVersion() {
        assertEquals(SemanticVersion.parse("v1.2"), SemanticVersion.parse("1.2.0"));
    }

    @Test
    void stableReleaseIsNewerThanPrerelease() {
        assertTrue(SemanticVersion.parse("1.0.0").compareTo(
                SemanticVersion.parse("1.0.0-SNAPSHOT")) > 0);
    }

    @Test
    void comparesNumericPrereleasePartsAsNumbers() {
        assertTrue(SemanticVersion.parse("1.0.0-beta.10").compareTo(
                SemanticVersion.parse("1.0.0-beta.2")) > 0);
    }

    @Test
    void ignoresBuildMetadata() {
        assertEquals(
                SemanticVersion.parse("1.2.3+build.1"),
                SemanticVersion.parse("1.2.3+build.2")
        );
    }
}

File: modules/api/src/test/kotlin/ru/privatenull/pnlibrary/api/version/SemanticVersionTest.kt
package ru.privatenull.pnlibrary.api.version

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class SemanticVersionTest {
    @Test
    fun `numeric prerelease identifiers compare numerically`() {
        assertTrue(SemanticVersion.parse("2.0.0-beta.10") > SemanticVersion.parse("2.0.0-beta.2"))
    }

    @Test
    fun `rejects invalid semver leading zeros and empty identifiers`() {
        assertNull(SemanticVersion.tryParse("02.0.0"))
        assertNull(SemanticVersion.tryParse("2.0.0-beta..1"))
        assertNull(SemanticVersion.tryParse("2.0.0-beta.01"))
        assertNull(SemanticVersion.tryParse("2.0.0+!"))
    }

    @Test
    fun `compares alphanumeric prerelease identifiers case sensitively`() {
        assertTrue(SemanticVersion.parse("1.0.0-BETA") < SemanticVersion.parse("1.0.0-beta"))
    }

    @Test
    fun `validates and ignores build metadata for precedence`() {
        assertEquals(
            SemanticVersion.parse("1.2.3+linux.x64"),
            SemanticVersion.parse("1.2.3+windows.x64"),
        )
    }

    @Test
    fun `rejects inverted range`() {
        assertThrows(IllegalArgumentException::class.java) { VersionRange.closed("3.0.0", "2.0.0") }
    }
}
